src/py21cmfast/drivers/param_config.py

```diff
@@ -172,63 +209,124 @@ def is_compatible_with(self, other: InputParameters) -> bool:
-                ip_args = {
-                    k.lstrip("_"): getattr(struct, k, None)
-                    for k in struct._inputs
-                    if k.lstrip("_") in [field.name for field in attrs.fields(cls)]
-                }
-                input_params.append(cls(**ip_args))
-
-        if len(input_params) == 0:
-            return cls(**defaults)
-        else:
-            # Combine all the parameter structs from input boxes
-            out = cls.combine(input_params)
-            # Now combine with provided kwargs
-            return attrs.evolve(out.merge(cls(**defaults)), redshift=redshift)
+        Does not change the input struct.
+        """
+        InputParameters.from_output_structs(
+            output_structs,
+            node_redshifts=self.node_redshifts,
+            cosmo_params=self.cosmo_params,
+            user_params=self.user_params,
+            astro_params=self.astro_params,
+            flag_options=self.flag_options,
+            random_seed=self.random_seed,
+        )
+
+    def evolve_inputstructs(self, **kwargs):
+        """Return an altered clone of the `InputParamters` structs.
+
+        Unlike clone(), this function takes fields from the constituent `InputStruct` classes
+        and only overwrites those sub-fields instead of the entire field
+        """
+        cosmo_params = self.cosmo_params.clone(
+            **{
+                k: v
+                for k, v in kwargs.items()
+                if k in self.cosmo_params.asdict().keys()
+            }
+        )
+        user_params = self.user_params.clone(
+            **{k: v for k, v in kwargs.items() if k in self.user_params.asdict().keys()}
+        )
+        astro_params = self.astro_params.clone(
+            **{
+                k: v
+                for k, v in kwargs.items()
+                if k in self.astro_params.asdict().keys()
+            }
+        )
+        flag_options = self.flag_options.clone(
+            **{
+                k: v
+                for k, v in kwargs.items()
+                if k in self.flag_options.asdict().keys()
+            }
+        )
+
+        return self.clone(
+            cosmo_params=cosmo_params,
+            user_params=user_params,
+            astro_params=astro_params,
+            flag_options=flag_options,
+        )
 
     @classmethod
-    def from_template(cls, name, random_seed):
-        """Construct full InputParameters instance from native or TOML file template."""
-        return cls(**create_params_from_template(name), random_seed=random_seed)
+    def from_template(cls, name, **kwargs):
+        """Construct full InputParameters instance from native or TOML file template.
+
+        Takes `InputStruct` fields as keyword arguments which overwrite the template/default fields
+        """
+        return cls(
+            **create_params_from_template(name),
+            **kwargs,
+        )
 
     @classmethod
     def from_inputstructs(
```

Comment:
again I prefer `from_input_structs`

Reply:
I think I can remove this method entirely now that we require every field